### PR TITLE
feat(nms): Equipment dashboard Federation Gateway Table

### LINKED ISSUE #2842: Add Federation Equipment dashboard and Gateway overview page

### PR DESCRIPTION
**Pull request by ato-dani**
## Summary
This is a draft PR because it is a stacked pr upon an existing open PR #7746 . Once, that PR is approved, this PR would be changed from draft to a normal PR. A federation gateway table was added to the equipment dashboard. It would display the basic configurations of the federation gateways. 

## Test Plan
A test was added under FEGEquipmentGatewayTest.js file to test how the federation gateway table is rendered. 

## Additional Information
Screen Shot 2021-06-28 at 1 51 09 PM

### CODE REVIEW

**Comment by uri200** [comment]
I think that table would make more sense from top to down, this way we can add more columns per each service.

Does each service reports it's health to the cloud? If so we could show the health for gx, gy, s6a . And the whole gatway.

**Comment by ato-dani** [comment]
> I think that table would make more sense from top to down, this way we can add more columns per each service.
> 
> Does each service reports it's health to the cloud? If so we could show the health for gx, gy, s6a . And the whole gatway.

@uri200 Those information would be displayed in the Gateway Detail Page. That would be my next milestone project. When you click the gateways, you should go to the details page, and there detailed information would be present so that we won't overload all info about the network just in one page.

**Comment by uri200** [comment]
> > I think that table would make more sense from top to down, this way we can add more columns per each service.
> > Does each service reports it's health to the cloud? If so we could show the health for gx, gy, s6a . And the whole gatway.
> 
> @uri200 Those information would be displayed in the Gateway Detail Page. That would be my next milestone project. When you click the gateways, you should go to the details page, and there detailed information would be present so that we won't overload all info about the network just in one page.

oh I see. I was missing the header in that picture, so I couldn't tell. Thanks for clarifying

**Review by themarwhal — APPROVED** [review]
hmm not sure why I'm requested as an owner, will look into this. But approving to unblock. Since karthik is off, I'll add andrei and hanna for review

File: nms/app/packages/magmalte/app/components/context/RefreshContext.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/components/context/RefreshContext.js]: this comment is redundant - the same info is obvious in the return statement right below.

File: nms/app/packages/magmalte/app/components/context/RefreshContext.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/components/context/RefreshContext.js]: String union types are a little strange to use. If somebody needs to update one of the values they would need to find and replace all occurrences.

```suggestion
const RefreshTypeEnum = Object.freeze({SUBSCRIBER: 'subscriber', GATEWAY: 'gateway', ...})
type RefreshType = $Values<typeof RefreshTypeEnum>;
```

https://stackoverflow.com/questions/287903/how-can-i-guarantee-that-my-enums-definition-doesnt-change-in-javascript
https://flow.org/en/docs/types/utilities/

Comment (andreilee) [reply nms/app/packages/magmalte/app/components/context/RefreshContext.js]: Where there's comparisons to `'feg_gateway'`, they can then be updated as well
```
if (props.type === RefreshTypeEnum.FEG_GATWAY) {
```

File: nms/app/packages/magmalte/app/components/context/RefreshContext.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/components/context/RefreshContext.js]: Why's this `$FlowIgnore` here? Add details to the comment at least

Comment (andreilee) [reply nms/app/packages/magmalte/app/components/context/RefreshContext.js]: Also add to the other `$FlowIgnore` annotations in the function

**Review by andreilee — CHANGES_REQUESTED** [review]

File: nms/app/packages/magmalte/app/state/feg/EquipmentState.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/state/feg/EquipmentState.js]: You should be able to do a direct truthiness check on `id` here. A zero-length `gateway_id` is nonsense anyways.

https://www.sitepoint.com/javascript-truthy-falsy/

File: nms/app/packages/magmalte/app/state/feg/EquipmentState.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/state/feg/EquipmentState.js]: Capitalize the error messages. These are read by NMS users and would be better as understandable sentences.

File: nms/app/packages/magmalte/app/state/feg/EquipmentState.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/state/feg/EquipmentState.js]: Update this to specify the type - not a list

Comment (andreilee) [reply nms/app/packages/magmalte/app/state/feg/EquipmentState.js]: Also either this specify that it throws an error when it fails to fetch gateways, or returns an empty set

File: nms/app/packages/magmalte/app/components/context/RefreshContext.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/components/context/RefreshContext.js]: `fegGateways || {}` shouldn't be necessary here. Getting the health is arbitrary if there aren't any `fegGateways`.

File: nms/app/packages/magmalte/app/state/feg/EquipmentState.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/state/feg/EquipmentState.js]: Specify the only `@param` as type `GatewayStateProps`, and add these helpful comments to the definition of `GatewayStateProps` above.

File: nms/app/packages/magmalte/app/state/feg/EquipmentState.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/state/feg/EquipmentState.js]: Either the whole function body should be wrapped in the same `try..catch`, or these two error messages should be different and provide some additional context. Right now it's just redundant.

File: nms/app/packages/magmalte/app/views/equipment/FEGGatewayTable.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/views/equipment/FEGGatewayTable.js:138]: We should probably display something different if there's no federation gateways. Say, if they just haven't been configured yet, the NMS would just appear broken. Defer to @jingtien if necessary.

File: nms/app/packages/magmalte/app/views/equipment/FEGGatewayTable.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/views/equipment/FEGGatewayTable.js]: Actually, these health status constants can be put into an enum as well.

File: nms/app/packages/magmalte/app/views/equipment/FEGGatewayTable.js
Comment (andreilee) [inline nms/app/packages/magmalte/app/views/equipment/FEGGatewayTable.js]: Include these into the health status enum as an unknown value type

**Review by andreilee — CHANGES_REQUESTED** [review]

File: nms/app/packages/magmalte/app/views/equipment/FEGGatewayTable.js
Comment (ato-dani) [reply nms/app/packages/magmalte/app/views/equipment/FEGGatewayTable.js:138]: I don't understand why it would appear broken. I am saying this because if there are no federation gateways in the network then all the headers of the table are there(name, id, gx ,...), but there is no data in there. Anyone seeing that would notice that there is no federation gateway to display under those headers. If you created a new test lte network go to the equipment page without adding any gateway, a blank table will be seen there. 
Screen Shot 2021-07-21 at 2 33 13 PM

I think anyone seeing this would notice that there are no gateways in the network(the same applies here too). Besides the count of 0 on table title will tell you that there are no federation gateways. Does that make sense?

Comment (andreilee) [reply nms/app/packages/magmalte/app/views/equipment/FEGGatewayTable.js:138]: Sounds good. Thanks for the detailed response.

**Review by andreilee — APPROVED** [review]

**Pull request merged**
